### PR TITLE
fix #46778, `precompile()` for abstract but compileable signatures

### LINKED ISSUE #46778: `precompile` fails to cache code for calls that are not fully specialized (non-concrete signature)
In https://github.com/JuliaLang/julia/pull/46690, I want to deserialize an object on a precompile worker and it is quite important that this is fully precompiled since a lot of precompile workers are spawned and each of these have a separate process and thus they would all need to compile the same code over and over if this isn't the case.

For arbitrary data types, `deserialize` has a function that takes `::DataType` and will not get specialized:

https://github.com/JuliaLang/julia/blob/aae8c484fd4ef9c9d7119e00f5eee679c905e542/stdlib/Serialization/src/Serialization.jl#L1468

As an example workload we can look at the following code

```julia
using Serialization

struct MyStruct
    x::String
end

s = MyStruct("foo")

io = IOBuffer()
serialize(io, s)
seekstart(io)
@time deserialize(io);
```

Running that with `--trace-compile=stderr` the following precompile statements are emitted from the `deserialize` call:

```julia
precompile(Tuple{typeof(Serialization.deserialize), Serialization.Serializer{Base.GenericIOBuffer{Array{UInt8, 1}}}, DataType})
```

However, restarting Julia and executing these before the `deserialize` call we can see that the `precompile` call for the `DataType` fails:

```julia
julia> precompile(Tuple{typeof(Serialization.deserialize), Serialization.Serializer{Base.IOStream}, DataType})
false
```

and the compilation time is still there.

However, if we specialize the function on the data type (as done in https://github.com/JuliaLang/julia/pull/46690/commits/6e0b2dad814b7a6603c47a5d0523591109e99dc0#diff-0fbb91f060d958d93fcf0a101ade4d87b365f3664a409276c591d7faf77acb67) the precompilation statements we get is:

```julia
precompile(Tuple{typeof(Serialization.deserialize), Serialization.Serializer{Base.GenericIOBuffer{Array{UInt8, 1}}}, Type{Main.MyStruct}})
```

the compilation time is then removed:

```julia
julia> @time deserialize(io);
  0.000014 seconds (14 allocations: 1.148 KiB)
```

It would be good if the code that is compiled for signatures that are non-concrete could be cached. They clearly are when you run the code since you don't have to compile it for every call.

### PR DESCRIPTION
**Pull request by JeffBezanson**
fix #46778

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -1141,7 +1141,7 @@ function const_prop_methodinstance_heuristic(interp::AbstractInterpreter,
     # was able to cut it down to something simple (inlineable in particular).
     # If so, there's a good chance we might be able to const prop all the way
     # through and learn something new.
-    if isdefined(method, :source) && is_inlineable(method.source)
+    if isdefined(method, :source) && method.source !== nothing && is_inlineable(method.source)
```
Comment (vtjnash) [inline base/compiler/abstractinterpretation.jl:1144]: seems like something is wrong if the `source` is defined but somehow also corrupted?

File: src/gf.c [hunk 2]
```diff
@@ -2274,36 +2307,52 @@ jl_method_instance_t *jl_get_specialization1(jl_tupletype_t *types JL_PROPAGATES
...
+    jl_method_match_t *match = NULL;
+    if (n == 1) {
+        match = (jl_method_match_t*)jl_array_ptr_ref(matches, 0);
+    }
+    else {
+        for (i = 0; i < n; i++) {
+            jl_method_match_t *match1 = (jl_method_match_t*)jl_array_ptr_ref(matches, i);
+            if (jl_isa_compileable_sig(match1->spec_types, match1->method)) {
+                match = match1;
+                break;
```
Comment (vtjnash) [inline src/gf.c:2347]: why only one match, chosen at random, and not all of them?

File: base/compiler/abstractinterpretation.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/compiler/abstractinterpretation.jl:1144]: Oh oops, I didn't intend to include this in this PR.

**Comment by KristofferC** [comment]
Bump! Good to go?

**Comment by JeffBezanson** [comment]
Oh, it looks like generate_precompile has an explicit check for this:
```
            ms = length(ps) == 1 ? Base._methods_by_ftype(ps[1], 1, Base.get_world_counter()) : Base.methods(ps...)
            ms isa Vector || continue
```

**Comment by KristofferC** [comment]
Oh...

**Pull request merged**